src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNodeFactory.cs

```diff
@@ -0,0 +1,421 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    using static AnalyzedNode;
+
+    // Create a common tree from expressions as well as patterns.
+    // Since the tree would be similar in both cases, we can later rewrite
+    // the whole thing as either a pattern or expression if needed.
+    internal static class AnalyzedNodeFactory
+    {
+        public static AnalyzedNode? Visit(IOperation operation)
+        {
+            return VisitCore(operation, input: null);
+        }
+
+        private static AnalyzedNode? VisitCore(IOperation operation, Evaluation? input)
+        {
+            Debug.Assert(input is null || operation is IPatternOperation || operation is IInvalidOperation);
+
+            return operation switch
+            {
+                IBinaryOperation op => VisitBinaryOperator(op),
+                IPatternOperation op => VisitPatternOperation(input, op),
+                IIsTypeOperation op => new Type(VisitInput(op.ValueOperand), op.TypeOperand, op.Syntax),
+                IIsPatternOperation op => VisitPatternOperation(VisitInput(op.Value), op.Pattern),
+                IPatternCaseClauseOperation op => VisitBinary(op.Pattern, op.Guard, disjunctive: false, input: null),
+                ISwitchExpressionArmOperation op => VisitBinary(op.Pattern, op.Guard, disjunctive: false, input: null),
+                IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } op => VisitNotOperator(op),
+                var op => VisitInput(op)
+            };
+        }
+
+        [return: NotNullIfNotNull("operation")]
+        private static Evaluation? VisitInput(IOperation? operation, Evaluation? input = null)
```

Comment:
consider making `input` non optional here.  being more explicit on hte caller side is helpful as it makes one have to consder if passing null makes sense.